R/scatter_plot_financial.R

```diff
@@ -124,19 +102,68 @@ prepare_scatter_plot_financial <- function(data, benchmarks_arg, scenario_arg, y
 #' @param mode A character vector.
 #' @param col A character vector.
 #'
-#' @return A numerical value.
+#' @return A vector.
 #' @noRd
-calculate_rank <- function(data, mode_arg, col) {
-  rank <- switch(mode_arg,
-    "equal_weight_finance" = mean(data[[col]], na.rm = TRUE),
-    "worst_case_finance" = {
-      data <- data[data[[mode_arg]] != 0, ]
-      mean(data[[col]], na.rm = TRUE)
+calculate_rank <- function(data, mode, col) {
+  rank <- switch(mode,
+    "equal_weight_finance" = {
+      rank <- ave(data[[col]], data[["bank_id"]],
+        FUN = function(x) mean(x, na.rm = TRUE)
+      )
     },
+    "worst_case_finance" = ,
     "best_case_finance" = {
-      data <- data[data[[mode_arg]] != 0, ]
-      mean(data[[col]], na.rm = TRUE)
+      data <- data[data[[mode]] != 0, ]
+      rank <- ave(data[[col]], data[["bank_id"]],
+        FUN = function(x) mean(x, na.rm = TRUE)
+      )
     }
   )
-  rank
+  list(rank, data)
+}
+
+#' Implement Rank
+#'
+#' @param data A data frame.
+#' @param mode A string.
+#'
+#' @return A data frame.
+#' @noRd
+calculate_scatter_plot_financial <- function(data, mode) {
+  data <- calculate_rank(data, mode, aka("profile_ranking"))[[2]]
+
+  data$emission_profile_average <- calculate_rank(data, mode, aka("profile_ranking"))[[1]]
+  data$transition_risk_average <- calculate_rank(data, mode, aka("transition_risk_score"))[[1]]
+
+  data
+}
+
+#' Plot the scatter financial plot
+#'
+#' @param data A data frame.
+#'
+#' @return A [ggplot] object.
+#' @noRd
+plot_scatter_financial <- function(data) {
+  emission_rank_legend <- label_emission_rank() |> format_label()
+  transition_risk_legend <- label_transition_risk() |> format_label()
+
+  emission_rank_plot <- ggplot(data, aes(x = .data$amount_total, color = .data$bank_id)) +
+    geom_point(aes(y = .data$emission_profile_average)) +
+    facet_wrap(~ .data$benchmark, scales = "fixed") +
+    ylim(0, NA) +
+    xlim(0, NA) +
+    ylab(emission_rank_legend) +
+    theme_tiltplot()
+
+  transition_score_plot <- ggplot(data, aes(x = .data$amount_total, color = .data$bank_id)) +
+    geom_point(aes(y = .data$transition_risk_average)) +
+    facet_wrap(~ .data$benchmark, scales = "fixed") +
+    ylim(0, NA) +
+    xlim(0, NA) +
+    ylab(transition_risk_legend) +
+    theme_tiltplot()
```

Comment:
I see some potential for reducing duplication. 

Maybe you can extract an internal helper with the common code. I think its interface may be something like `plot_scatter_financial_impl(data, type = c("emission_profile", "transition_risk")`. 


```r
plot_scatter_financial <- function(data) {
  ...
  ggarrange(
    plot_scatter_financial_impl(data, type = "emission_profile"),
    plot_scatter_financial_impl(data, type = "transition_risk")  
  )
}
```

The `type` argument would be used to construct the name of the column that needs to be used inside `geom_point()`, and also `type` would be used to construct the call to either `emission_profile_legend()` or `transition_risk_legend()` (via `get()`).

Here is the gist of my suggestion:

``` r
# These are functions you already have 
emission_profile_legend <- function() "I'm emission_profile's legend"
transition_risk_legend <- function() "I'm transition_risk's legend"

type <- "emission_profile"

col <- paste0(type, "_average")
# Now you can use this in `geom_point()`
col
#> [1] "emission_profile_average"
f_legend <- get(paste0(type, "_legend"))

# Now you can use this in `ylab()`
f_legend()
#> [1] "I'm emission_profile's legend"
```

Let me  know if this is confusing. I'm not feeling very eloquent today :-(

Reply:
It is very clear, thank you !